protocol/proto/decode.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+
+	"github.com/m3db/m3x/pool"
+)
+
+type decoder struct {
+	r          *bufio.Reader
+	sizeBuffer []byte
+	dataBuffer []byte
+	bytesPool  pool.BytesPool
+}
+
+// NewDecoder decodes a new decoder.
+func NewDecoder(r io.Reader, opts EncodeDecoderOptions) Decoder {
+	return newDecoder(r, opts)
+}
+
+func newDecoder(r io.Reader, opts EncodeDecoderOptions) *decoder {
+	if opts == nil {
+		opts = NewEncodeDecoderOptions()
+	}
+	pool := opts.BytesPool()
+	return &decoder{
+		r:          bufio.NewReaderSize(r, opts.BufferSize()),
+		sizeBuffer: getByteSliceWithLength(sizeBufferSize, pool),
+		bytesPool:  pool,
+	}
+}
+
+func (d *decoder) Decode(m Unmarshaler) error {
+	size, err := d.decodeSize()
+	if err != nil {
+		return err
+	}
+	return d.decodeData(m, size)
+}
+
+func (d *decoder) decodeSize() (int, error) {
+	if _, err := io.ReadFull(d.r, d.sizeBuffer); err != nil {
+		return 0, err
+	}
+	size := binary.BigEndian.Uint32(d.sizeBuffer)
```

Comment:
Use a shared var for endianness during encoding/decoding?

Reply:
sounds good